File: java/04_015.java
public interface NewSystem {
    void processData();
}

public class OldSystem {
    void processLegacyData() { /* old logic */ }
}

public class SystemAdapter implements NewSystem {
    private OldSystem oldSystem;
    
    public void processData() {
        oldSystem.processLegacyData();
    }
}

File: data structures and algorithms/cpp/04_015.cpp
#include <memory>

class NewSystem {
public:
    virtual ~NewSystem() = default;
    virtual void processData() = 0;
};

class OldSystem {
public:
    void processLegacyData() {
        // old logic
    }
};

class SystemAdapter : public NewSystem {
private:
    std::unique_ptr<OldSystem> oldSystem;
    
public:
    SystemAdapter(std::unique_ptr<OldSystem> system) 
        : oldSystem(std::move(system)) {}
    
    void processData() override {
        oldSystem->processLegacyData();
    }
};